components/search/README.md

```diff
@@ -0,0 +1,68 @@
+# Search
+
+The **Search Rust component** provides various [search service](https://firefox-source-docs.mozilla.org/toolkit/search/index.html) functions for Firefox Desktop, Android and iOS.
+
+It is currently under construction and not yet used.
+
+## Architecture
+
+This module currently has one function, `filter_engine_configuration`, which receives configurations from remote settings and details of the user's environment. It returns a list of search engines to be presented to the user.
+
+This component follows the architecture of the other [Application Services Rust components](https://mozilla.github.io/application-services/book/index.html): a cross-platform Rust core, and platform-specific bindings for Firefox Desktop, Android, and iOS. These bindings are generated automatically using the [UniFFI](https://mozilla.github.io/uniffi-rs/) tool.
+
+### For contributors
+
+This section is a primer for engineers contributing code to the Search Rust component.
+
+`search.udl` describes the component's interface for foreign language consumers. UniFFI uses this file to generate the language bindings for each platform.
+
+`store.rs` contains the implementation of `SuggestStore` and most of the Suggest component's tests.
+
+## Documentation
+
+Each Rust file contains [inline documentation](https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html) for types, traits, functions, and methods.
+
+Documentation for `pub` symbols is written with application developers in mind: even if you're a Desktop, Android or iOS developer, the Rust documentation is meant to give you an understanding of how the Gecko, Kotlin and Swift bindings work.
+
+You can see the documentation for all public symbols by running from the command line:
+
+```shell
+cargo doc --open
+```
+
+By convention, symbols that are exported as part of the component's foreign language interface have `pub` visibility, and symbols that are only used within the component have `pub(crate)` or private visibility. As an exception to this convention, symbols with [doctests](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html) are also `pub`, because doctests [can only link against public symbols](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html#include-items-only-when-collecting-doctests).
```

Comment:
I'm feeling a little on the fence about this convention. I experimented with it in Suggest, but not sure if the awkwardness of typing `pub(crate)` vs. `pub` each time is worth it. I think @bendk had some good thoughts about it also!

When I was experimenting with this convention, I thought that having Rust docs would be better than nothing for application developers. Even if the names and syntax didn't quite match up, at least the `cargo doc` output could be somewhat helpful for explaining a component's core concepts and abstractions to Desktop, Android, and iOS devs.

I was worried that using `pub` would clutter up the `cargo doc` output with Rust-only symbols, which are super helpful if you're working on the component itself, but not so much if you just want to consume it.

I think I overestimated how many Android / iOS devs would be open to installing Rust and running `cargo doc`, though 😅 I get the feeling that our consumers (completely understandably!) just want to see docs in Android Studio and Xcode, and "run `cargo doc` then figure out how to translate its output to the Kotlin / Swift API you're using" might not be super interesting to them.

Reply:
I'm not sure if this qualifies as a good thought, but another way of controlling visibility is to make everything `pub` if it should be visible outside the module, and use `pub use` vs `use`  in the top-level module to distinguish what gets exposed to consumer crates.  I kind of like that model, since there's one place that controls what's exposed.  This has worked out okay for me, but I'm not totally sold on it.  I keep feeling like I'll find a corner case that I hadn't though out.

Reply:
Thanks, I'll think I'll drop that text for now, and go with the `pub use` option for the time being.